### ISSUE TITLE
Performance regression: WebUtility.UrlDecode 2.5x slower on AMD only

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/corefx#39762]
For some reason, the [System.Net.Tests.Perf_WebUtility.Decode_DecodingRequired](
https://github.com/dotnet/performance/blob/7d96fd3dea66b83a0b02222e44e02e762a990c32/src/benchmarks/micro/corefx/System.Runtime.Extensions/Perf.WebUtility.cs#L13-L14) benchmark have regressed 2.5x on AMD only.

## Repro

```log
git clone https://github.com/dotnet/performance.git
cd performance
# if you don't have cli installed and want python script to download the latest cli for you
py .\scripts\benchmarks_ci.py -f netcoreapp2.2 netcoreapp3.0 --filter System.Net.Tests.Perf_WebUtility.Decode_DecodingRequired
# if you do
dotnet run -p .\src\benchmarks\micro\MicroBenchmarks.csproj -c Release -f netcoreapp2.2 --runtimes netcoreapp2.2 netcoreapp3.0 --filter System.Net.Tests.Perf_WebUtility.Decode_DecodingRequired
```

## System.Net.Tests.Perf_WebUtility.Decode_DecodingRequired

| Conclusion |   Base |   Diff | Base/Diff | Modality | Operating System     | Arch  | Processor Name                              | Base Runtime    | Diff Runtime                     |
| ---------- | ------:| ------:| ---------:| --------:| -------------------- | ----- | ------------------------------------------- | --------------- | --------------------------------- |
| Same       | 182.30 | 182.19 |      1.00 |          | ubuntu 18.04         | 64bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Same       | 154.89 | 153.16 |      1.01 |          | Windows 10.0.18362   | 64bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Same       | 309.98 | 305.69 |      1.01 |          | ubuntu 16.04         | 64bit | Intel Xeon CPU E5-2673 v4 2.30GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-01|
| Same       | 320.10 | 331.77 |      0.96 |          | ubuntu 18.04         | 64bit | Intel Xeon CPU E5-2673 v4 2.30GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-01|
| Same       | 211.93 | 210.11 |      1.01 |          | macOS Mojave 10.14.5 | 64bit | Intel Core i7-5557U CPU 3.10GHz (Broadwell) | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Same       | 182.27 | 180.18 |      1.01 |          | Windows 10.0.18362   | 64bit | Intel Core i7-5557U CPU 3.10GHz (Broadwell) | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Faster     | 138.34 | 130.48 |      1.06 |          | Windows 10.0.18362   | 64bit | Intel Core i7-7700 CPU 3.60GHz (Kaby Lake)  | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     | 175.79 | 420.39 |      0.42 |          | Windows 10.0.18362   | 64bit | AMD Ryzen 7 1800X                           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Slower     | 134.67 | 327.60 |      0.41 |          | Windows 10.0.18362   | 64bit | AMD Ryzen 9 3900X                           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-28373-12|
| Same       | 182.93 | 173.77 |      1.05 |          | Windows 10.0.18362   | 32bit | Intel Xeon CPU E5-1650 v4 3.60GHz           | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-27919-09|
| Same       | 205.06 | 203.17 |      1.01 |          | Windows 10.0.18362   | 32bit | Intel Core i7-5557U CPU 3.10GHz (Broadwell) | .NET Core 2.2.6 | .NET Core 3.0.0-preview8-28373-17|

/cc @danmosemsft @tannergooding @billwert @DrewScoggins

**Comment by karelz** [comment]
@wtgodbe will you be able to take a look?

**Comment by wtgodbe** [comment]
Sure thing

**Comment by wtgodbe** [comment]
@adamsitnik do you have an estimate of when this regression started? When was the last time you ran these benchmarks?

**Comment by karelz** [comment]
@davidsh do you think this is blocker for 3.0? I have a bit doubts.
It would be nice to understand what caused it (@wtgodbe will look at that), but it is IMO just "nice", not necessary ...

Thoughts?

**Comment by adamsitnik** [comment]
Please keep in mind that the regression affects only AMD hardware. I suspect that some of the methods that `UrlDecode` uses internally have regressed and this is the outcome of it.

I currently don't have access to AMD machine, so please excuse me for "guessing" instead of providing an answer using a profiler.

From what I can see this method uses `Encoding.Utf8.GetBytes()` to get the utf8 representation, then does some simple operations and calls `Array.Copy` :

https://github.com/dotnet/corefx/blob/e23f83e6172ef2d879342e1c4883012749b91ef1/src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs#L180

https://github.com/dotnet/corefx/blob/e23f83e6172ef2d879342e1c4883012749b91ef1/src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs#L186

https://github.com/dotnet/corefx/blob/e23f83e6172ef2d879342e1c4883012749b91ef1/src/System.Web.HttpUtility/src/System/Web/HttpUtility.cs#L201

https://github.com/dotnet/corefx/blob/e72c6c53baddb02a630e8e6592a59d89e30b856b/src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs#L248-L291

Most probably it's hitting https://github.com/dotnet/coreclr/issues/25505 which affects `Array.Copy`

@tannergooding is there any chance that you could re-run the `Decode_DecodingRequired` benchmark and see if https://github.com/dotnet/coreclr/pull/25763 helped?

```cmd
--filter System.Net.Tests.Perf_WebUtility.Decode_DecodingRequired
```

@karelz It should definitely not be a blocker for 3.0.

**Comment by tannergooding** [comment]
Yes, I can try to find some time this week (feel free to ping me if I forget 😄).

**Comment by wtgodbe** [comment]
I'll try now with an SDK containing the coreclr fix

**Comment by davidsh** [comment]
>@davidsh do you think this is blocker for 3.0? I have a bit doubts.
It would be nice to understand what caused it (@wtgodbe will look at that), but it is IMO just "nice", not necessary ...
Thoughts?

The networking stack (HTTP etc.) doesn't use System.Web.HttpUtility.  It uses System.Uri for "UrlEncode/Decode" operations. So, I don't think this is 3.0 blocker.

**Comment by wtgodbe** [comment]
Never mind, it just occured to me that I don't have a machine with an AMD processor 😆. @tannergooding do you?

**Comment by tannergooding** [comment]
Yes, three of them (two personal, one work). If it isn't high priority though, I will be getting to it in a few days

**Comment by wtgodbe** [comment]
Sounds good, I'll mark this for 5.0
